File: backend/08-review/src/main/kotlin/com/dokandar/review/data/DbBootstrap.kt
package com.dokandar.review.data

import com.dokandar.review.Config
import com.dokandar.review.observability.Log
import java.io.File
import java.sql.DriverManager
import java.sql.SQLException

// Self-bootstrap: validate db name → create-if-missing (catch 42P04) → apply migrations — before bind.
object DbBootstrap {
    fun ensure() {
        require(Config.pgDb.matches(Regex("^[A-Za-z_][A-Za-z0-9_]*$"))) { "refusing unsafe db name: ${Config.pgDb}" }
        DriverManager.getConnection(Config.pgUrl("postgres"), Config.pgUser, Config.pgPassword).use { admin ->
            val rs = admin.prepareStatement("SELECT 1 FROM pg_database WHERE datname=?").apply { setString(1, Config.pgDb) }.executeQuery()
            if (!rs.next()) {
                try { admin.createStatement().execute("CREATE DATABASE \"${Config.pgDb}\""); Log.info("review.boot", "created database ${Config.pgDb}") }
                catch (e: SQLException) { if (e.sqlState != "42P04") throw e }
            }
        }
        DriverManager.getConnection(Config.pgUrl(), Config.pgUser, Config.pgPassword).use { c ->
            val dir = listOf("migrations", "/app/migrations").map { File(it) }.firstOrNull { it.isDirectory }
                ?: throw IllegalStateException("migrations dir not found")
            (dir.listFiles { f -> f.name.endsWith(".sql") } ?: emptyArray()).sortedBy { it.name }.forEach { f ->
                c.createStatement().execute(f.readText())
            }
            Log.info("review.boot", "migrations applied; schema ready")
        }
    }
}

File: backend/07-coupon/src/Data/DbBootstrap.cs
using System.Text.RegularExpressions;
using Npgsql;
using Coupon.Observability;

namespace Coupon.Data;

// Self-bootstrap: create the DB if missing, then apply migrations/*.sql — BEFORE the listener binds.
public static class DbBootstrap
{
    public static async Task EnsureAsync()
    {
        if (!Regex.IsMatch(Config.PgDb, "^[A-Za-z_][A-Za-z0-9_]*$"))
            throw new InvalidOperationException($"refusing unsafe db name: {Config.PgDb}");

        await using (var admin = new NpgsqlConnection(Config.PgConn("postgres")))
        {
            await admin.OpenAsync();
            bool exists;
            await using (var check = new NpgsqlCommand("SELECT 1 FROM pg_database WHERE datname=@d", admin))
            {
                check.Parameters.AddWithValue("d", Config.PgDb);
                exists = await check.ExecuteScalarAsync() != null;
            }
            if (!exists)
            {
                try
                {
                    await using var create = new NpgsqlCommand($"CREATE DATABASE \"{Config.PgDb}\"", admin);
                    await create.ExecuteNonQueryAsync();
                    Log.Info("coupon.boot", $"created database {Config.PgDb}");
                }
                catch (PostgresException e) when (e.SqlState == "42P04") { /* race: already exists */ }
            }
        }

        await using (var conn = new NpgsqlConnection(Config.PgConn()))
        {
            await conn.OpenAsync();
            var dir = Directory.Exists("migrations") ? "migrations" : "/app/migrations";
            foreach (var f in Directory.GetFiles(dir, "*.sql").OrderBy(x => x))
            {
                var sql = await File.ReadAllTextAsync(f);
                await using var cmd = new NpgsqlCommand(sql, conn);
                await cmd.ExecuteNonQueryAsync();
            }
            Log.Info("coupon.boot", "migrations applied; schema ready");
        }
    }
}
